File: Days 001 - 010/Day 05/GettingValuesFromPairs.cs
using System;

namespace DailyCodingProblems
{
	internal class Day5
	{
		private delegate int PairFunction(int a, int b);
		private delegate int ReturnedFunction(PairFunction f);

		private static int Main(string[] args)
		{
			Console.WriteLine(Car(Cons(3, 4)));
			Console.WriteLine(Cdr(Cons(3, 4)));

			Console.ReadLine();

			return 0;
		}

		private static ReturnedFunction Cons(int a, int b)
		{
			int Pair(PairFunction f) => f(a, b);

			return Pair;
		}

		private static int Car(ReturnedFunction pair)
		{
			int First(int a, int b) => a;

			return pair(First);
		}

		private static int Cdr(ReturnedFunction pair)
		{
			int Second(int a, int b) => b;

			return pair(Second);
		}
	}
}

File: Days 001 - 010/Day 05/GettingValuesFromPairs.cpp
#include <functional>
#include <iostream>

using PairFunction = std::function<int(int, int)>;
using ReturnedFunction = std::function<int(PairFunction)>;

ReturnedFunction cons(int a, int b)
{
	auto pair = [a, b](PairFunction f) -> int
	{
		return f(a, b);
	};

	return pair;
}

int car(ReturnedFunction pair)
{
	auto firstVar = [](int a, int b) -> int
	{
		return a;
	};

	return pair(firstVar);
}

int cdr(ReturnedFunction pair)
{
	auto secondVar = [](int a, int b) -> int
	{
		return b;
	};

	return pair(secondVar);
}

int main(int argc, char* argv[])
{
	std::cout << car(cons(3, 4)) << "\n";
	std::cout << cdr(cons(3, 4)) << "\n";

	std::cin.get();

	return 0;
}